File: app/src/main/java/com/koko/www/androiddemo/useInterface/layout/recyclerView/RecyclerView_Activity.java
package com.koko.www.androiddemo.useInterface.layout.recyclerView;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.koko.www.androiddemo.R;

public class RecyclerView_Activity extends AppCompatActivity {

    private RecyclerView mRecyclerView;
    private GridLayoutManager mLayoutManager;
    private MyRecyclerViewAdapter myRecyclerViewAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.layout_recyclerview);
        mRecyclerView = findViewById(R.id.my_recycler_view);

        // in content do not change the layout size of the RecyclerView
        mRecyclerView.setHasFixedSize(true);
        // use a linear layout manager
        //spanCount = 3：表示一横行显示3列
        mLayoutManager = new GridLayoutManager(this, 4);
        mLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            //表示该position的Item，占用该横行的几列
            @Override
            public int getSpanSize(int position) {
                return 1;
            }
        });
        mRecyclerView.setLayoutManager(mLayoutManager);

        myRecyclerViewAdapter = new MyRecyclerViewAdapter();
        mRecyclerView.setAdapter(myRecyclerViewAdapter);
    }
}

File: app/src/main/java/com/koko/www/androiddemo/useInterface/recyclerView/RecyclerViewActivity.kt
package com.koko.www.androiddemo.useInterface.recyclerView

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.recyclerview.widget.DividerItemDecoration

import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.koko.www.androiddemo.R

class RecyclerViewActivity : AppCompatActivity() {

    private var mRecyclerView: RecyclerView? = null
    private var mLayoutManager: LinearLayoutManager? = null
    private var myRecyclerViewAdapter: ListAdapter? = null
    private var myMultiItemAdapter: MultiItemAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.layout_recyclerview)
        mRecyclerView = findViewById(R.id.my_recycler_view)

        initAdapter()

        upDate()

    }

    private fun initAdapter() {
        // in content do not change the layout size of the RecyclerView
        mRecyclerView!!.setHasFixedSize(true)
        mLayoutManager = LinearLayoutManager(this)
        mRecyclerView!!.layoutManager = mLayoutManager

        //添加item分界线
        val decoration = DividerItemDecoration(this, DividerItemDecoration.VERTICAL)
        mRecyclerView!!.addItemDecoration(decoration)

        //使用ListAdapter
//        myRecyclerViewAdapter = ListAdapter()
        //使用多种布局Adapter
        myMultiItemAdapter = MultiItemAdapter()

        mRecyclerView!!.adapter = myMultiItemAdapter
    }

    private fun upDate() {
        myRecyclerViewAdapter?.notifyDataSetChanged()
    }
}
